### ISSUE TITLE
Generic non-location format for time zones unavailable?

### ISSUE THREAD

**Issue by MenoData** [ThreeTen/threetenbp#44]
I tried:

```Java
  DateTimeFormatter zf =
    DateTimeFormatter.ofPattern("zzzz", Locale.ENGLISH).withZone(ZoneId.of("America/Los_Angeles"));
  System.out.println(LocalDateTime.now().format(zf)); // Pacific Time
```

This code compiles in Java-8 (v1.8.0_77), is runnable and yields the expected result ("Pacific Time"). However, using ThreetenBP fails with following exception:

```Java
Exception in thread "main" org.threeten.bp.temporal.UnsupportedTemporalTypeException: Unsupported field: InstantSeconds
	at org.threeten.bp.LocalDate.get0(LocalDate.java:593)
	at org.threeten.bp.LocalDate.getLong(LocalDate.java:572)
	at org.threeten.bp.LocalDateTime.getLong(LocalDateTime.java:628)
	at org.threeten.bp.format.DateTimePrintContext$1.getLong(DateTimePrintContext.java:181)
	at org.threeten.bp.format.DateTimeFormatterBuilder$ZoneTextPrinterParser.print(DateTimeFormatterBuilder.java:3370)
	at org.threeten.bp.format.DateTimeFormatterBuilder$CompositePrinterParser.print(DateTimeFormatterBuilder.java:1995)
	at org.threeten.bp.format.DateTimeFormatter.formatTo(DateTimeFormatter.java:1385)
	at org.threeten.bp.format.DateTimeFormatter.format(DateTimeFormatter.java:1359)
	at org.threeten.bp.chrono.ChronoLocalDateTime.format(ChronoLocalDateTime.java:263)
	at org.threeten.bp.LocalDateTime.format(LocalDateTime.java:1828)
```

I remember our discussion on [stackoverflow](https://stackoverflow.com/questions/36722873/datetimeformatter-not-compatible-with-simpledateformat). Following questions:

- Is the observed exception a bug or a feature of ThreetenBP?
- Are there any plans to update ThreetenBP to newer subversions of Java-8 (higher than 8u20)?
- If a solution for given code above is not possible then how can I achieve the generic non-location name of a time zone in a different way?

**Comment by jodastephen** [comment]
ThreeTenBP is a best efforts project to get JSR-310 on earlier JDKs. Not everything is perfect, but I don't have the time to make it perfect. The key problem here is that JDKs before JDK 8 will not contain the localized "generic non-location" data, so the whole question is moot. (Basically, it is not ThreeTenBP's job to provide _data_ in JDK 6 and 7 that was added in 8.

Since ThreeTenBP is for JDK 6 and 7, the question about Java 8u20 is unclear.

**Comment by MenoData** [comment]
Thank you very much for your response which addresses my third question. I will try to clarify my other two questions.

The same exception I observed also happens with following code (similar stacktrace):

```
System.out.println(
  ZoneId.of("America/Los_Angeles").getDisplayName(TextStyle.FULL, Locale.ENGLISH));
```

Even if the generic name data are not available on Java 6+7, I would at least expect this code to be runnable and to fall back maybe to "specific standard time", IANA-id or something else. The exception should IMHO simply not occur. The root cause of the exception is that the ThreetenBP-code expects to find the missing field "INSTANT_SECONDS" without any alternative.

About Java 8u20, I have looked at your commit history, and the last release [synchronizing ThreetenBP with Java-8](http://www.threeten.org/threetenbp/changes-report.html#a0.9) affected 8u20. But Java-8 is in active development. So my second question means, do you still have time or plans to synchronize ThreetenBP with newer bugfixes in Java-8? If yes, then maybe this problem can at least be partially solved (avoiding the exception, but with different data).

Thanks in advance.

**Comment by jodastephen** [comment]
I would be happy to see the codebase updated to match the behviour of later Java 8 releases. However, practically speaking I don't have time to do that (I'm performing minimal maintenance on the project). The error you point out might well be one that could be better handled.

**Comment by sschaap** [comment]
As @jodastephen points out, JDK 6 and 7 do not provide localized resources for a generic timezone name. That said, the javadoc for [TimeZone#getDisplayName(Locale)](https://docs.oracle.com/javase/6/docs/api/java/util/TimeZone.html#getDisplayName()) states that, if daylight-savings yes/no is not specified, the name of the standard time (i.e. no daylight savings) is returned. The same logic could be applied to zone name formatting and, consequently, ZoneId#getDisplayName().

As far as I can tell, this would require a relatively trivial patch in ZoneTextPrinterParser#print(DateTimePrintContext, StringBuilder) replacing

```
Long epochSec = context.getTemporal().getLong(INSTANT_SECONDS);
Instant instant;
if (epochSec != null) {
    instant = Instant.ofEpochSecond(epochSec);
} else {
    instant = Instant.ofEpochSecond(-200L * 365 * 86400);  // about 1770
}
TimeZone tz = TimeZone.getTimeZone(zone.getId());
boolean daylight = zone.getRules().isDaylightSavings(instant);
```

with something along the lines of (untested):

```
TemporalAccessor temporal = context.getTemporal();
TimeZone tz = TimeZone.getTimeZone(zone.getId());
boolean daylight;
if (temporal.isSupported(INSTANT_SECONDS)) {
    long epochSec = temporal.getLong(INSTANT_SECONDS);
    Instant instant = Instant.ofEpochSecond(epochSec);
    daylight = zone.getRules().isDaylightSavings(instant);
} else {
    daylight = false
}
```

This has the side effect of removing the `// about 1170` case. However, this code seems unreachable in any case as TemporalAccessor#getLong(TemporalField) is declared to return a primitive long, which means that `Long epochSec` is always non-null. Not sure why that's in there.

TL;DR Issue seems trivial to patch. Mainly needs someone to write a patch set to match Java 8 behavior more closely.
